core/src/shred.rs

```diff
@@ -0,0 +1,426 @@
+        }
+        slice_len
+    }
+}
+
+pub struct Shredder {
+    slot: u64,
+    index: u32,
+    parent: Option<u64>,
+    _fec_ratio: u64,
+    signer: Arc<Keypair>,
+    pub shreds: Shreds,
+    pub active_shred: Option<SignedShred>,
+    pub active_offset: usize,
+}
+
+impl Write for Shredder {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let mut current_shred = self
+            .active_shred
+            .take()
+            .or_else(|| {
+                Some(
+                    self.parent
+                        .take()
+                        .map(|parent| {
+                            // If parent slot is provided, assume it's first shred in slot
+                            SignedShred::new(Shred::FirstInSlot(self.new_first_shred(parent)))
+                        })
+                        .unwrap_or_else(||
+                            // If parent slot is not provided, and since there's no existing shred,
+                            // assume it's first shred in FEC block
+                            SignedShred::new(Shred::FirstInFECSet(self.new_data_shred()))),
+                )
+            })
+            .unwrap();
+
+        let written = self.active_offset;
+        let slice_len = match current_shred.shred.borrow_mut() {
+            Shred::FirstInSlot(s) => s.write_at(written, buf),
+            Shred::FirstInFECSet(s) => s.write_at(written, buf),
+            Shred::Data(s) => s.write_at(written, buf),
+            Shred::LastInFECSetData(s) => s.write_at(written, buf),
+            Shred::LastInSlotData(s) => s.write_at(written, buf),
+            Shred::Coding(s) => s.write_at(written, buf),
+            Shred::LastInFECSetCoding(s) => s.write_at(written, buf),
+            Shred::LastInSlotCoding(s) => s.write_at(written, buf),
+        };
+
+        let active_shred = if buf.len() > slice_len {
+            self.finalize_shred(current_shred);
+            // Continue generating more data shreds.
+            // If the caller decides to finalize the FEC block or Slot, the data shred will
+            // morph into appropriate shred accordingly
+            SignedShred::new(Shred::Data(DataShred::default()))
+        } else {
+            self.active_offset += slice_len;
+            current_shred
+        };
+
+        self.active_shred = Some(active_shred);
+
+        Ok(slice_len)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        if self.active_shred.is_none() {
+            return Ok(());
+        }
+        let current_shred = self.active_shred.take().unwrap();
+        self.finalize_shred(current_shred);
+        Ok(())
+    }
+}
+
+impl Shredder {
+    pub fn new(slot: u64, parent: Option<u64>, fec_ratio: u64, signer: &Arc<Keypair>) -> Self {
+        Shredder {
+            slot,
+            index: 0xff,
+            parent,
+            _fec_ratio: fec_ratio,
+            signer: signer.clone(),
```

Comment:
see if you can remove this clone and the ones from `sign` below

Reply:
Removed from `sign()`